### DISCUSSION TITLE
Cara menghubungkan shiny dengan postgresSQL

### DISCUSSION THREAD

**Discussion by rafiff23** [teamalgoritma/community#672] [Q&A] [answered]
https://github.com/teamalgoritma/community/discussions/621 dari diskusi disini mas @Yusraf002 sudah menjelaskan bagaimana caranya menghubgkan postgressql dengan R, kemudian codingan saya seperti ini 
```
require(RPostgreSQL)
library(shiny)
library(shinymanager)
library(DBI)
library(glue)
library(RPostgres)

con <- dbConnect(
  Postgres(),
  user = "username",
  password = "password",
  host = "localhost",
  port = 5432,
  dbname = "database_name"
)

# Create a table in the database
dbExecute(con, "CREATE TABLE credentials (id SERIAL, user TEXT, password TEXT, admin BOOLEAN)")

# Insert data into the table
dbExecute(con, "INSERT INTO credentials (user, password, admin) VALUES ('user', 'user', FALSE), ('Adtree', '270792', TRUE)")


ui <- fluidPage(
  tags$h2("My secure application"),
  verbatimTextOutput("auth_output")
)
ui <- secure_app(ui)


server <- function(input, output, session) {
  res_auth <- secure_server(
    check_credentials = check_credentials(
      con,
      table_name = "credentials",
      username_col = "user",
      password_col = "password",
      salt = "somesalt"
    )
  ) 
  auth_output <- reactive({
    reactiveValuesToList(res_auth)
  })
  
  # access info
  observe({
    print(auth_output())
  })
}

shinyApp(ui, server)
```

dengan error seperti 
```
Error: could not connect to server: Connection refused (0x0000274D/10061)
	Is the server running on host "localhost" (::1) and accepting
	TCP/IP connections on port 5432?
could not connect to server: Connection refused (0x0000274D/10061)
	Is the server running on host "localhost" (127.0.0.1) and accepting
	TCP/IP connections on port 5432?
```

apakah saya harus mendownload PostgresSQL dulu di windows atau membuat akun di PostgresSQL jika iya apakah ada step2 nya?
dan apakah ini bisa berjalan di aplikasi shiny?

**Comment by jafarhf1** [comment]
Dear Pak @rafiff23 

Terima kasih Pak untuk pertanyaannya, akan kami coba lihat terlebih dahulu dan mencarikan solusinya

**Comment by Yusraf002** [comment]
Dear @rafiff23 ,
Terimakasih atas pertanyaannya pak. Betul sekali pak, sebelum Bapak dapat menggunakan PostgreSQL perlu untuk mengunduh dan menginstal di perangkat yang Bapak gunakan terlebih dahulu. untuk tahapannya kurang lebih seperti berikut:

1. Unduh installer PostgreSQL dari situs resmi [PostgreSQL](https://www.postgresql.org)
2. Pilih versi yang sesuai dengan sistem operasi yang bapak gunakan (32-bit atau 64-bit).
3. Setelah selesai mengunduh, buka installer PostgreSQL.
4. Ikuti petunjuk pada layar untuk menginstal PostgreSQL.

untuk lebih jelasnya bapak bisa mengikuti panduan dan mempelajari terkait PostgeSQL dari link berikut: 
[PostgreSQL (1) Installation and Create Your First Database ( SQL shell / PgAdmin4 )](https://youtu.be/KuQUNHCeKCk)

untuk PostgreSQl dapat berjalan di aplikasi Shiny dengan menggunakan paket `RPostgreSQL` dan `RPostgres` untuk menghubungkan ke server PostgreSQL dan menjalankan operasi database. Untuk menjalankannya Bapak bisa melakukan installasi package terlebih dahulu.

Sekian informasi dari kami, jika dirasa membantu dapat klik "mark as answer" pada comment ini. Terima kasih

Best Regards,
Algoritma Team.

**Comment by rafiff23** [comment]
![image](https://github.com/teamalgoritma/community/assets/90491715/6018f21a-4646-4ff6-b43d-84c75611df5f)

Setelah ini saya harus ngapain ya mas @Yusraf002 ? saya udah create database baru bernama credentials tapi saya tidak tau bagaimana caranya menghubungkan database ini ke R dan dimana caranya tau username  dan password nya untuk fungsi "con"

**Reply by Yusraf002** [reply, marked as answer]
Halo @rafiff23, untuk username bisa bapak temukan pada tab `Properties` sedangkan untuk passwordnya menggunakan password yang bapak buat saat pertama install
<img width="960" alt="username postgre" src="https://github.com/teamalgoritma/community/assets/127075678/71eac197-523e-4d16-ae0b-6087f5fd0ac4">

sehingga pada kode yang sudah bapak buat bisa ganti pada poin-poin berikut:
```{r}
con <- dbConnect(
  Postgres(),
  user = "username",  <- pada bagian ini bapak bisa isikan sesuai nama usernya
  password = "password", <- pada bagian password bapak isikan password yg bapak buat diawal saat instalasi
  host = "localhost",
  port = 5432,
  dbname = "database_name" <- disini bapak bisa isikan nama database yg sudah bapak buat
)
```

Sekian informasi dari kami, jika dirasa membantu dapat klik "mark as answer" pada comment ini. Terima kasih

Best Regards,
Algoritma Team.
